tests/poly/runner/generate_test_cases.py

```diff
@@ -0,0 +1,222 @@
+import argparse
+import sys
+
+import sympy
+import toml
+
+HEADER = """// WARNING: this file is autogenerated. Do not edit manually, instead see
+// tests/poly/runner/generate_test_cases.py
+
+//-------------------------------------------------------
+// entry and check_prefix are re-set per test execution
+// DEFINE: %{entry} =
+// DEFINE: %{check_prefix} =
+
+// DEFINE: %{compile} = heir-opt %s --heir-polynomial-to-llvm
+// DEFINE: %{run} = mlir-cpu-runner -e %{entry} -entry-point-result=void --shared-libs="%mlir_lib_dir/libmlir_c_runner_utils%shlibext,%mlir_runner_utils"
+// DEFINE: %{check} = FileCheck %s --check-prefix=%{check_prefix}
+//-------------------------------------------------------
+
+func.func private @printMemrefI32(memref<*xi32>) attributes { llvm.emit_c_interface }
+
+"""
+
+
+def test_setup(i: int):
+  entry_def = '// REDEFINE: %{entry} = test_' + str(i)
+  check_def = '// REDEFINE: %{check_prefix} = CHECK_TEST_' + str(i)
+  run = '// RUN: %{compile} | %{run} | %{check}'
+  return '\n'.join([entry_def, check_def, run])
+
+
+# Params:
+#  - ideal
+#  - cmod
+#  - p0
+#  - p1
+#  - test_number
+#  - gen_tensor_op
+#  - degree
+#  - expected
+#  - coefficient_list
+TEST_TEMPLATE = """
+#ideal_{{test_number}} = #poly.polynomial<{{ideal}}>
+#ring_{{test_number}} = #poly.ring<cmod={{cmod}}, ideal=#ideal_{{test_number}}>
+!poly_ty_{{test_number}} = !poly.poly<#ring_{{test_number}}>
+
+func.func @test_{{test_number}}() {{{{
+  %const0 = arith.constant 0 : index
+  %0 = poly.constant <{{p0}}> : !poly_ty_{{test_number}}
+  %1 = poly.constant <{{p1}}> : !poly_ty_{{test_number}}
+  %2 = poly.mul(%0, %1) : !poly_ty_{{test_number}}
+
+{gen_tensor_op}
+  %ref = bufferization.to_memref %tensor : memref<{{degree}}xi32>
+  %U = memref.cast %ref : memref<{{degree}}xi32> to memref<*xi32>
+  func.call @printMemrefI32(%U) : (memref<*xi32>) -> ()
+  return
+}}}}
+// expected_result: {{expected}}
+// CHECK_TEST_{{test_number}}: {{coefficient_list}}
+"""
+
+GEN_TENSOR_OP_TEMPLATE_WITH_TRUNC = """
+  %3 = poly.to_tensor %2 : !poly_ty_{{test_number}} -> tensor<{{degree}}x{{container_type}}>
+  %tensor = arith.{trunc_ext_op} %3 : tensor<{{degree}}x{{container_type}}> to tensor<{{degree}}xi32>
```

Comment:
interesting, I'm guessing this is because of the limitation of the print call being i32 that we have to check the result mod 2^32? I think that'll end up fine- if something was wrong, then chances are there's a case where it'll show up wrong mod 2^32.

could you add a quick comment for this template that we check equality mod 2^32 because of the function call limitation and expect to see issues in lowerings arise below i32? we can reevaluate if there's something funky we do for larger bit widths.